### ISSUE TITLE
[BUG] set artifact endpoint fails when overriding default step definitions

### ISSUE THREAD

**Issue by sgsabbage** [marklogic/marklogic-data-hub#5314]
#### The issue

Short description of the problem:

It looks like the set artifact endpoint is designed to allow a project to override default step definitions in [these lines of code](https://github.com/marklogic/marklogic-data-hub/blob/master/marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/stepDefinition.sjs#L62-L64)

However, this fails when deploying a new default step definition: with

>  Local message: failed to POST at /data-hub/5/data-services/artifacts/setArtifact.sjs: Internal Server Error. Server Message: JS-JAVASCRIPT: dir = dir + mutableArtifact.toLowerCase() + "/" + "/marklogic/"; -- Error running JavaScript request: TypeError: mutableArtifact.toLowerCase is not a function . See the MarkLogic server error log for further detail.

I suspect this should be `dir = dir + mutableArtifact.type.toLowerCase() + "/marklogic/";` - however I did want to confirm first that this was one of the intended uses for this endpoint? (As an aside, I am happy to submit PRs but need to check about the CLA internally)

What are the steps to reproduce the issue:
1) Create a new default step definition, e.g. `$PROJECT_DIR/step-definitions/matching/default-matching/default-matching.step.json`
2) Run the `hubDeployUserArtifacts` gradle task

#### Tech details

Which Operating System are you using?
Docker Container on WSL2, Windows 10

Which version of MarkLogic are you using?
10

Which version of the Data Hub are you using?
5.4

Log attached [debug.log](https://github.com/marklogic/marklogic-data-hub/files/6075673/debug.log)

**Comment by briantang** [comment]
Thanks for filing this @sgsabbage. The issue has been confirmed a bug and we are tracking it internally with https://project.marklogic.com/jira/browse/DHFPROD-7009. We plan to address it ASAP as a high priority bug and have it fixed for Data Hub 5.4.1.

**Comment by rjrudin** [comment]
@iowusu Does your project has custom step definitions in it with names starting with "default-"?

**Comment by rjrudin** [comment]
@sgsabbage Are you trying to override the OOTB default-matching step definition, or are you just looking to create your own?

**Comment by rjrudin** [comment]
The default step definition can't be overridden via a normal deploy. DHF constructs its URI as "/step-definition/(type)/marklogic/(name).step.json". When a step def is deployed, DHF constructs its URI as "/step-definition/(type)/(name)/(name).step.json". 

What use case do you have for overridding it?

We're going to release DHF 5.4.1 soon to fix this bug. @sgsabbage  @iowusu  In the mean time, the workaround is not to have any user-defined step definitions that start with "default-". Once 5.4.1 is released, your step definition names can start with "default-".

**Comment by iowusu** [comment]
Yes, the project has default-matching and default-mastering.

**Comment by rjrudin** [comment]
Thanks @iowusu - what's your use case for overriding those? Are you swapping in your own modulePath? I don't think DHF has ever intended for the OOTB step definitions to be overridden, so curious why you need to do that.

**Comment by iowusu** [comment]
```
{
  "name" : "memberTransaction",
  "description" : "",
  "batchSize" : 100,
  "threadCount" : 4,
  "stopOnError" : false,
  "options" : { },
  "version" : 0,
  "steps" : {
    "1" : {
      "name" : "memberTransactionIngestion",
      "description" : "",
      "options" : {
        "additionalCollections" : [ ],
        "sourceQuery" : "cts.collectionQuery([])",
        "targetEntity" : "",
        "collections" : [ ],
        "sourceDatabase" : "RDL",
        "sourceCollection" : "",
        "outputFormat" : "json",
        "targetDatabase" : "RDL"
      },
      "customHook" : null,
      "retryLimit" : 0,
      "batchSize" : 0,
      "threadCount" : 0,
      "stepDefinitionName" : "memberTransactionIngestion",
      "stepDefinitionType" : "CUSTOM"
    },
    "2" : {
      "name" : "memberTransactionHarmonize",
      "description" : "",
      "options" : {
        "additionalCollections" : [ ],
        "sourceQuery" : "cts.collectionQuery([\"recipient\"])",
        "targetEntity" : "",
        "collectorDatabase" : "RDL",
        "collections" : [ ],
        "sourceDatabase" : "RDL",
        "sourceCollection" : "",
        "outputFormat" : "json",
        "targetDatabase" : "TDS"
      },
      "customHook" : null,
      "retryLimit" : 0,
      "batchSize" : 0,
      "threadCount" : 0,
      "stepDefinitionName" : "memberTransactionHarmonize",
      "stepDefinitionType" : "CUSTOM"
    },
    "3" : {
      "name" : "memberTransactionMatching",
      "description" : "",
      "options" : {
        "stepUpdate" : false,
        "additionalCollections" : [ ],
        "sourceQuery" : "cts.collectionQuery([\"memberTransactions\"])",
        "acceptsBatch" : true,
        "targetEntity" : "member",
        "sourceDatabase" : "TDS",
        "collections" : [ "memberTransactionMatching", "member" ],
        "permissions" : "data-hub-operator,read,data-hub-operator,update",
        "matchOptions" : {
          "propertyDefs" : {
            "property" : [ {
              "localname" : "firstName",
              "name" : "firstName"
            } ]
          },
          "algorithms" : {
            "algorithm" : [ {
              "name" : "double-metaphone",
              "function" : "double-metaphone",
              "namespace" : "http://marklogic.com/smart-mastering/algorithms",
              "at" : "/com.marklogic.smart-mastering/algorithms/double-metaphone.xqy"
            }, {
              "name" : "thesaurus",
              "function" : "thesaurus",
              "namespace" : "http://marklogic.com/smart-mastering/algorithms",
              "at" : "/com.marklogic.smart-mastering/algorithms/thesaurus.xqy"
            }, {
              "name" : "zip-match",
              "function" : "zip-match",
              "namespace" : "http://marklogic.com/smart-mastering/algorithms",
              "at" : "/com.marklogic.smart-mastering/algorithms/zip.xqy"
            }, {
              "name" : "standard-reduction",
              "function" : "standard-reduction"
            } ]
          },
          "collections" : {
            "content" : [ ]
          },
          "scoring" : {
            "add" : [ {
              "propertyName" : "firstName",
              "weight" : "10"
            } ],
            "expand" : [ ],
            "reduce" : [ ]
          },
          "actions" : {
            "action" : [ ]
          },
          "thresholds" : {
            "threshold" : [ {
              "above" : "8",
              "label" : "firstName",
              "action" : "merge"
            } ]
          },
          "tuning" : {
            "maxScan" : 200
          }
        },
        "sourceCollection" : "memberTransactions",
        "outputFormat" : "json",
        "targetDatabase" : "TDS"
      },
      "customHook" : {
        "module" : "",
        "parameters" : { },
        "user" : "",
        "runBefore" : false
      },
      "retryLimit" : 0,
      "batchSize" : 100,
      "threadCount" : 4,
      "stepDefinitionName" : "default-matching",
      "stepDefinitionType" : "MATCHING"
    }
  }
}
```
This is the initial flow file that has default merge.

**Comment by rjrudin** [comment]
That's a flow with steps; those won't cause the error. The error is coming from something in your ./step-definitions directory.

**Comment by iowusu** [comment]
The ./step-definitions directory does not contain "default-".  I do think "default-" is causing my error.

**Comment by sgsabbage** [comment]
> The default step definition can't be overridden via a normal deploy. DHF constructs its URI as "/step-definition/(type)/marklogic/(name).step.json". When a step def is deployed, DHF constructs its URI as "/step-definition/(type)/(name)/(name).step.json".

Coming back to this, as far as I can tell, the code that's throwing the error is designed to deploy any step definitions starting with `default-` to the path `/step-definition/(type)/marklogic/` (it's just that the calculation of `type` is incorrect) : https://github.com/marklogic/marklogic-data-hub/blob/c4e11d15c0a2b2c4c57a2c77b43fcbd9c77e17d3/marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/stepDefinition.sjs#L62-L64

I just wanted to highlight that because it sounds like this behaviour may not be what's intended from the code?

**Comment by rjrudin** [comment]
Our analysis indicated that the scenario this code was intended to support - writing OOTB DHF artifacts - never involves executing this code; the OOTB DHF step definitions are instead loaded directly from the /v1/documents endpoint. So the fix for this is to simply remove the "if" condition altogether, which removes the buggy line of code as well. The net effect is that a user can now create a step definition with a name starting with "default-". 

However, this is not intended to allow a user to override OOTB DHF step definitions, which I don't believe has ever been a supported nor intended requirement. That's because when DHF loads its own step definitions, it constructs a URI of "/step-definition/(type)/marklogic/(name).step.json". Whereas a user step definition receives a URI of "/step-definition/(type)/(name)/(name).step.json".

**Comment by kkanthet** [comment]
Issues fixed in 5.4.1 which is now GA.
